### PR TITLE
pd: bi-streaming heartbeat

### PR DESCRIPTION
**Pull request by BusyJay**
This implementation assumes that pd server closes connection on error.

### CODE REVIEW

**Comment by BusyJay** [comment]
Depend on pingcap/kvproto#152

File: src/raftstore/store/worker/pd.rs [hunk 1]
```diff
@@ -273,12 +243,53 @@ impl<T: PdClient> Runner<T> {
...
+        handle.spawn(f);
+        self.is_hb_receiver_scheduled = true;
+    }
 }
 
 impl<T: PdClient> Runnable<Task> for Runner<T> {
     fn run(&mut self, task: Task, handle: &Handle) {
         debug!("executing task {}", task);
 
+        if !self.is_hb_receiver_scheduled {
```
Comment (siddontang) [inline src/raftstore/store/worker/pd.rs]: Later, in most cases, PD will not return the response if the region has nothing to change. Is it ok to check scheduler here?

Comment (BusyJay) [inline src/raftstore/store/worker/pd.rs]: I don't see any problem here. The check ensure heartbeat receiver is hooked only once.

File: src/pd/async/util.rs [hunk 2]
```diff
@@ -102,6 +164,12 @@ impl LeaderClient {
 
         {
             let mut inner = self.inner.wl();
+            let (tx, rx) = client.region_heartbeat();
+            inner.hb_sender = Either::Left(Some(tx));
+            if let Either::Right(ref mut task) = inner.hb_receiver {
+                task.unpark();
```
Comment (overvenus) [inline src/pd/async/util.rs]: Why does it need `unpark` here?

Comment (BusyJay) [inline src/pd/async/util.rs]: If hb_receiver is a Task, L83 has been called, call `unpark` to notify the receiver poller.

**Comment by overvenus** [comment]
Now, pingcap/kvproto#152 is up to date.

**Comment by BusyJay** [comment]
PTAL

File: src/pd/async/client.rs [hunk 3]
```diff
@@ -173,8 +174,8 @@ impl PdClient for RpcClient {
         req.set_header(self.header());
         req.set_region_id(region_id);
 
-        let executor = |client: &PDClient, req: pdpb::GetRegionByIDRequest| {
-            let handler = client.get_region_by_id_async(req);
+        let executor = |client: &RwLock<Inner>, req: pdpb::GetRegionByIDRequest| {
+            let handler = client.read().unwrap().client.get_region_by_id_async(req);
```
Comment (siddontang) [inline src/pd/async/client.rs]: can use HandyRwLock here?

File: src/pd/async/client.rs [hunk 4]
```diff
@@ -206,28 +207,52 @@ impl PdClient for RpcClient {
...
-        let executor = |client: &PDClient, req: pdpb::RegionHeartbeatRequest| {
-            let handler = client.region_heartbeat_async(req);
-            handler.map_err(Error::Grpc)
-                .and_then(|resp| {
-                    try!(check_resp_header(resp.get_header()));
-                    Ok(resp)
-                })
-                .boxed()
+        let executor = |client: &RwLock<Inner>, req: pdpb::RegionHeartbeatRequest| {
+            let mut inner = client.write().unwrap();
```
Comment (siddontang) [inline src/pd/async/client.rs]: `wl()` here?

**Comment by BusyJay** [comment]
Do we need to handle metrics PD_SEND_MSG_HISTOGRAM in region heartbeat? @siddontang @overvenus

**Comment by siddontang** [comment]
@BusyJay region heartbeat is a bi-stream now, I don't know how to get its duration. Maybe we should only use counter for it. 

Btw, please fix conflict and we can go on. 

PTAL @overvenus @disksing

**Comment by overvenus** [comment]
How about attach a timer in  [`Request`](https://github.com/pingcap/tikv/pull/1888/files#diff-d0a7d28df9e505e1859adea84587afd0R185)? Although, additional time will be recorded too, like building futures, waiting for being scheduled.

**Comment by siddontang** [comment]
any update?

**Comment by BusyJay** [comment]
There are some changes after merging master, PTAL

File: src/pd/async/client.rs (hunk 4 above)
Comment (siddontang) [inline src/pd/async/client.rs]: any update?

**Comment by disksing** [comment]
@siddontang @overvenus PTAL

**Comment by overvenus** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**